### PR TITLE
Time series buffer

### PR DESCRIPTION
**Pull request by bangxiangyong**
Should resolve https://github.com/Met4FoF/agentMET4FOF/issues/129

Added MetrologicalAgentBuffer class which utilises TimeSeriesBuffer for modelling metrological data:
https://github.com/Met4FoF/agentMET4FOF/blob/time_series_buffer/agentMET4FOF/metrological_agents.py

See running example: 
https://github.com/Met4FoF/agentMET4FOF/blob/time_series_buffer/agentMET4FOF_tutorials/buffering/metrological_buffering.py

### CODE REVIEW

File: agentMET4FOF/metrological_agents.py [hunk 1]
```diff
@@ -192,3 +193,118 @@ def custom_plot_function(self, data, sender_agent, **kwargs):
...
+        message_from = message["from"]
+        message_data = message["data"]
+
+        # check if sender agent has sent any message before:
+        # if it did,then append, otherwise create new entry for the input agent
+        if message_from not in self.buffer:
+            self.update(message_from, message_data)
+            return 0
+        else:
+            self._concatenate(iterable=self.buffer[message_from],data=message_data, concat_axis=concat_axis)
```
Comment (BjoernLudwigPTB) [inline agentMET4FOF/metrological_agents.py:293]: I am pretty sure we do not need this. After overwriting `_concatenate` the base `store` should work, I think.

**Review by BjoernLudwigPTB — APPROVED** [review]
Awesome work, Xiang! I added a bunch of type hints and docstrings and tweak minor bits of algorithmic approaches in agents.py. Besides I am prette sure we do not need to overwrite `store` in metrological agents. Let's check that...

File: agentMET4FOF/metrological_agents.py (hunk 1 above)
Comment (BjoernLudwigPTB) [reply agentMET4FOF/metrological_agents.py:293]: @bangxiangyong Do you see a reason, to leave it in?

**Comment by bangxiangyong** [comment]
There appears to be a minor code conflict in the imports, do you know which ones should we use ? @BjoernLudwigPTB

File: agentMET4FOF/metrological_agents.py (hunk 1 above)
Comment (BjoernLudwigPTB) [reply agentMET4FOF/metrological_agents.py:293]: @bangxiangyong Before we merge this, do you see a reason to leave this store in? I think it is redundant and not really needed.

Comment (BjoernLudwigPTB) [reply agentMET4FOF/metrological_agents.py:293]: @bangxiangyong I think we can drop the overwriding of the `store` method. Do you agree?

**Comment by BjoernLudwigPTB** [comment]
> There appears to be a minor code conflict in the imports, do you know which ones should we use ? @BjoernLudwigPTB

This is resolved.

**Comment by BjoernLudwigPTB** [comment]
@bangxiangyong the conflicts mentioned above are actually very minor. The worst is, that after merging `develop` which works almost fine, since these confilcts above are really easy to resolve, the output of the monitor agents look very strange. This branch and `develop` seem to have diverged to far, to apply your changes as they are. I noticed that in this version here the `buffer_size` parameter of the monitor agents does not seem to be used properly meaning the graphs go on and on and datapoints are endlessly added. It is used in the `develop` branch though and excessive datapoints from the beginning of the `buffer` are dropped as new points come in. Please check your implementation after doing `git merge develop` and resolving the conflicts (which are not the cause of any unexpected behaviour). If you like, I can resolve the conflicts as well, but with all the changes from `develop` this will "break" the plots. Just let me know... ideally early before the sprint review.

**Comment by BjoernLudwigPTB** [comment]
We noticed that the introduced tutorial _metrological buffering_ only shows an empty plot. We will resolve this in the reopened issue #129 and a new PR.

**Pull request merged**